### ISSUE TITLE
Walking repo has unexpected behavior with ignored files

### ISSUE THREAD

**Issue by bk2204** [git-lfs/git-lfs#3428]
Currently, when we walk a repo using our repo walker, we have some differences from Git with respect to ignored files:

- If a file is checked in but matches a pattern in `.gitignore`, we treat it as ignored.
- We don't honor negative patterns in `.gitignore` (see #3190 and #3409).

We should maintain the same behavior as Git in this case, possibly by using `git ls-files` if necessary.

**Comment by jerick1976** [comment]
We have hit this issue within our environment when a repository has an extremely large set of files to traverse or when there are a large number of .gitignore files to process.  The issue is reproducible.  We hit it in 2.5.1 and 2.7.1.  Are there any tests you would like me to run?

**Comment by bk2204** [comment]
I don't think there's any tests that need to be run in particular. We're aware of the issue, but nobody has had time to work on it yet.

**Comment by jerick1976** [comment]
I believe https://github.com/git-lfs/git-lfs/issues/3668 may be a duplicate of this issue.

**Comment by bk2204** [comment]
Yes, I believe they're both related.  I'm working on a solution that should speed things up and address both issues.

**Comment by bk2204** [comment]
Would you be willing to try the branch on my remote named `ls-filters-file`?  That branch should provide a `git ls-files`-based implementation, which should be significantly faster than the current approach.

**Comment by jerick1976** [comment]
I believe I have some willing Canaries.  I'll pull it into our dev environment and let you know.

**Comment by jerick1976** [comment]
Greetings,

The shared patch version of git-lfs is also giving the same results for component commit slowness.  In this situation there were three files between 25 and 30 MB pushed to LFS.  What else do you need from me test wise?


$ git lfs version
git-lfs/2.8.0 (unknown; linux amd64; go 1.13.1; git 6b0f4bd5)

$ time git push https://dscm-write.cisco.com/git/153/comp-15302059.git testing_git_commit testing_git_commit/7 refs/notes/*
Uploading LFS objects: 100% (8230/8230), 221 GB | 0 B/s, done.
Enumerating objects: 11, done.
Counting objects: 100% (11/11), done.
Delta compression using up to 6 threads
Compressing objects: 100% (6/6), done.
Writing objects: 100% (6/6), 752 bytes | 188.00 KiB/s, done.
Total 6 (delta 2), reused 0 (delta 0)
To https://dscm-write.cisco.com/git/153/comp-15302059.git
2879948ed..e8608723b testing_git_commit -> testing_git_commit
* [new tag] testing_git_commit/7 -> testing_git_commit/7

real 22m56.100s
user 18m26.954s
sys 4m31.464s

**Comment by bk2204** [comment]
We've merged a similar patch (not mine) already, so at this point, you're welcome to test and see if that improves things for you.  Additionally, if you're not using locking, setting `lfs.setlockablereadonly` to `false` will likely improve performance.

I should point out that your example shows that you uploaded 8230 files worth 221 GB, which is likely the majority of your 22 minute time here.

**Comment by jerick1976** [comment]
Beyond a performance change what is the impact of setting this value to false?
